### ISSUE TITLE
Low results on SemanticKITTI validation set

### ISSUE THREAD

**Issue by PinTang1999** [SCPNet/Codes-for-SCPNet#8]
Thank you for your excellent paper and the priliminary code. However,  it's seems that the reproduced results is much lower than ones you claimed in your paper. The best results I obtained is completion IoU of 50.3 and semantic mIoU of 22.792. And I found that the completion sub-network and segmentation sub-network in your code is already given. Then, I should obtain the results: 48.5 (completion IoU) and 34.4 (semantic mIoU) as shown in Table 8 and Table 6(b) in your paper. But the reproduced mIoU (22.792) is much lower than your published result (34.4) with a gap of almost 12. I wonder if there are some key points that I've missed and could you help give me get the correct results? 
![image](https://github.com/SCPNet/Codes-for-SCPNet/assets/66257799/6f21771d-e6b3-4acd-841f-813394812b61)

**Comment by anhquancao** [comment]
Hello @PinTang1999 , I saw the author release a [checkpoint here](https://github.com/SCPNet/Codes-for-SCPNet/issues/2#issuecomment-1629979572). I am wondering if you were able to obtain good results with this checkpoint.

**Comment by PinTang1999** [comment]
> 

Thank you. I've seen that but I didn't try it. The authors do not provide any information about this checkpoint (like if it is trained with distillation code). Hence, I think it's meaniningless to try it. And I just trained the model using the released code, which performs much worse than the authors claimed.

**Comment by PinTang1999** [comment]
> Hello @PinTang1999 , I saw the author release a [checkpoint here](https://github.com/SCPNet/Codes-for-SCPNet/issues/2#issuecomment-1629979572). I am wondering if you were able to obtain good results with this checkpoint.

Thank you. I've seen that but I didn't try it. The authors do not provide any information about this checkpoint (like if it is trained with distillation code). Hence, I think it's meaniningless to try it. And I just trained the model using the released code, which performs much worse than the authors claimed.

**Comment by glimmer-shining** [comment]
I got the same problem and I have try the checkpoint on the validation set with 37.5 miou,   but I  cant not get the same 34.4 miou   
as the paper claimed

**Comment by anhquancao** [comment]
Hi @PinTang1999, I digged in the author code and I realized that the current training script doesn't using the label rectification. This could be the problem.

**Comment by PinTang1999** [comment]
> Hi @PinTang1999, @glimmer-shining , I digged in the author code and I realized that the current training script doesn't using the label rectification. It is also the multiscans version.

The 34.4 mIoU in the paper doesn't use the proposed label rectification, which is the same as the released code. But the released code can not reproduce this result, but only obtain 22.792 mIoU. That's the problem.

**Comment by anhquancao** [comment]
Thank you for the clarification! It's true that the 34.4 doesn't account for label rectification. Also, it's odd that the provided checkpoint is from epoch 3, especially when the paper states the model underwent 30 epochs of training.

**Comment by anhquancao** [comment]
After completing 30 training epochs with a batch size of 2 on an A100, I also achieved significantly lower numbers than those reported in the paper.

```
Validation per class iou: 
IoU class 1 [car] = 0.440
IoU class 2 [bicycle] = 0.119
IoU class 3 [motorcycle] = 0.090
IoU class 4 [truck] = 0.201
IoU class 5 [other-vehicle] = 0.080
IoU class 6 [person] = 0.025
IoU class 7 [bicyclist] = 0.031
IoU class 8 [motorcyclist] = 0.000
IoU class 9 [road] = 0.642
IoU class 10 [parking] = 0.210
IoU class 11 [sidewalk] = 0.411
IoU class 12 [other-ground] = 0.005
IoU class 13 [building] = 0.312
IoU class 14 [fence] = 0.175
IoU class 15 [vegetation] = 0.365
IoU class 16 [trunk] = 0.209
IoU class 17 [terrain] = 0.365
IoU class 18 [pole] = 0.225
IoU class 19 [traffic-sign] = 0.053
Current val completion iou is 0.474
Current val miou is 20.843 while the best val miou is 21.501
Current val loss is 3.073

```

**Comment by anhquancao** [comment]
Hello,

I reached out to the authors, and they kindly provided some suggestions, which I believe will be beneficial to all of you. Here is their response:

> Thanks for the detailed response. It seems that the reproduced completion IoU (47.4) is close to the reported value (48.5). The biggest performance gap lies in the semantic mIoU, i.e., 21.5 v.s. 34.4. This suggests that the network can roughly accomplish the class-agnostic completion task but performs badly in class-level recognition. The followings are some suggestions:
> 
> 1) Check the version of other important dependencies:
> torch-scatter, python, Numpy-indexed, numba, etc
> 
> 2) Adopt the two-stage training strategy. For example, you may first load the pre-trained weight of the segmentation model from the Cylinder3D repo (https://github.com/xinge008/Cylinder3D) or the PVKD repo (https://github.com/cardwing/Codes-for-PVKD). Then, you can freeze the segmentation network and only train the completion sub-network to see if the performance of the semantic recognition can be improved. An alternative solution is to first train the segmentation sub-network and then freeze the segmentation sub-network and only train the completion sub-network if we do not use the pre-trained model.
> 
> 3) Since the class imbalance problem is severe in the SemanticKITTI benchmark, you may adopt some strategies to mitigate the problem, e.g., reweight the cross entropy loss using the class frequency vector, perform stronger data augmentations such as LaserMix and PolarMix, carefully select a learning rate or learning rate schedule, resampling data with rare classes, use the pseudo labels on the training set to serve as the auxiliary soft supervision signals, use the decoupling head trick to divide classes into different groups and let the model learn each group separately, etc.
>

**Comment by PinTang1999** [comment]
> Hello @glimmer-shining @PinTang1999 ,
> 
> I reached out to the authors, and they kindly provided some suggestions, which I believe will be beneficial to all of you. Here is their response:
> 
> > Thanks for the detailed response. It seems that the reproduced completion IoU (47.4) is close to the reported value (48.5). The biggest performance gap lies in the semantic mIoU, i.e., 21.5 v.s. 34.4. This suggests that the network can roughly accomplish the class-agnostic completion task but performs badly in class-level recognition. The followings are some suggestions:
> > 
> > 1. Check the version of other important dependencies:
> >    torch-scatter, python, Numpy-indexed, numba, etc
> > 2. Adopt the two-stage training strategy. For example, you may first load the pre-trained weight of the segmentation model from the Cylinder3D repo (https://github.com/xinge008/Cylinder3D) or the PVKD repo (https://github.com/cardwing/Codes-for-PVKD). Then, you can freeze the segmentation network and only train the completion sub-network to see if the performance of the semantic recognition can be improved. An alternative solution is to first train the segmentation sub-network and then freeze the segmentation sub-network and only train the completion sub-network if we do not use the pre-trained model.
> > 3. Since the class imbalance problem is severe in the SemanticKITTI benchmark, you may adopt some strategies to mitigate the problem, e.g., reweight the cross entropy loss using the class frequency vector, perform stronger data augmentations such as LaserMix and PolarMix, carefully select a learning rate or learning rate schedule, resampling data with rare classes, use the pseudo labels on the training set to serve as the auxiliary soft supervision signals, use the decoupling head trick to divide classes into different groups and let the model learn each group separately, etc.

Spconv 1.x and 2.x do not result in the gap in point cloud semantic segmentation (if you train it on SemanticKITTI segmentation task). Hence, I don't think the version of spconv can lead to 10 mIoU drop. If we do not take suggestion 2 and 3 into consideration (as these tricks are not the contribution of SCPNet), I think we can make a conclusion that the proposed SCPNet without label rectification, multi-frame distillation and other tricks can only achieves about 20 mIoU on SemanticKITTI.
